python/private/attr_builders.bzl

```diff
@@ -0,0 +1,1284 @@
+        if cfg == "target" or cfg == None:
+            self.set_target()
+        elif cfg == "exec":
+            self.set_exec()
+        elif cfg == "none":
+            self.set_none()
+        else:
+            self.set_implementation(cfg)
+    if "exec_group" in kwargs:
+        self.set_exec(kwargs.pop("exec_group"))
+
+    if "implementation" in kwargs:
+        self.set_implementation(kwargs.pop("implementation"))
+
+    return self
+
+def _AttrCfg_from_attr_kwargs_pop(attr_kwargs):
+    """Creates a `AttrCfg` from the cfg arg passed to an attribute bulider.
+
+    Args:
+        attr_kwargs: dict of attr kwargs, it's "cfg" key will be removed.
+
+    Returns:
+        {type}`AttrCfg`
+    """
+    cfg = attr_kwargs.pop("cfg", None)
+    if not types.is_dict(cfg):
+        kwargs = {"cfg": cfg}
+    else:
+        kwargs = cfg
+    return _AttrCfg_new(**kwargs)
+
+def _AttrCfg_implementation(self):
+    """Tells the custom transition function, if any and applicable.
+
+    Returns:
+        {type}`callable | None` the custom transition function to use, if
+        any, or `None` if a different config mode is being used.
+    """
+    return self._state["value"] if self._state["which"] == "impl" else None
+
+def _AttrCfg_none(self):
+    """Tells if none cfg (`config.none()`) is set.
+
+    Returns:
+        {type}`bool` True if none cfg is set, False if not.
+    """
+    return self._state["value"] if self._state["which"] == "none" else False
+
+def _AttrCfg_target(self):
+    """Tells if target cfg is set.
+
+    Returns:
+        {type}`bool` True if target cfg is set, False if not.
+    """
+    return self._state["value"] if self._state["which"] == "target" else False
+
+def _AttrCfg_exec_group(self):
+    """Tells the exec group to use if an exec transition is being used.
+
+    Args:
+        self: implicitly added.
+
+    Returns:
+        {type}`str | None` the name of the exec group to use if any,
+        or `None` if `which_cfg` isn't `exec`
+    """
+    return self._state["value"] if self._state["which"] == "exec" else None
+
+def _AttrCfg_set_implementation(self, impl):
+    """Sets a custom transition function to use.
+
+    Args:
+        self: implicitly added.
+        impl: {type}`callable` a transition implementation function.
+    """
+    self._state["which"] = "impl"
+    self._state["value"] = impl
+
+def _AttrCfg_set_none(self):
+    """Sets to use the "none" transition."""
+    self._state["which"] = "none"
+    self._state["value"] = True
+
+def _AttrCfg_set_exec(self, exec_group = None):
+    """Sets to use an exec transition.
+
+    Args:
+        self: implicitly added.
+        exec_group: {type}`str | None` the exec group name to use, if any.
+    """
+    self._state["which"] = "exec"
+    self._state["value"] = exec_group
+
+def _AttrCfg_set_target(self):
+    """Sets to use the target transition."""
+    self._state["which"] = "target"
+    self._state["value"] = True
```

Comment:
nit: the `_state` keys should be constants in this file. Much easier to avoid typos.

Reply:
Will do.

Reply:
Done.